base/math.jl

```diff
@@ -70,11 +71,23 @@ end
 
 # evaluate p[1] + x * (p[2] + x * (....)), i.e. a polynomial via Horner's rule
 macro horner(x, p...)
+    @gensym val
     ex = esc(p[end])
     for i = length(p)-1:-1:1
-        ex = :(muladd(t, $ex, $(esc(p[i]))))
+        ex = :(muladd($val, $ex, $(esc(p[i]))))
     end
-    Expr(:block, :(t = $(esc(x))), ex)
+    Expr(:block, :($val = $(esc(x))), ex)
+end
+
+# evaluate p[1] + x * (p[2] + x * (....)), i.e. a polynomial via Horner's rule
+# and convert coefficients to same type as x
+macro horner_oftype(x, p...)
```

Comment:
I agree that it will generally change the degree of the polynomial. That being said, for example one case where it doesn't really change is if you also want to have native Float16 support (although on CPUs this is useless) the dispatch on the same polynomial (the one for  `::Float32`) works well in practice (from many cases on different function I found it's really hard to find good coefficients by reducing the degree from the Float32 version for Float16). For example this function will work just as well within accuracy for `Float16`.  Thus it's helpful to define this macro to take care of such dispatch cases.